### ISSUE TITLE
Introduce Search bar and view

### ISSUE THREAD

**Issue by MaxYablochkin** [flutter/flutter#117483]
#91605

<img width="1024" alt="unnamed" src="https://user-images.githubusercontent.com/102767277/208982037-b4ff9ef9-0342-406b-a8d9-89275e5ddf05.png">

Specs: https://m3.material.io/components/search/overview

**Comment by darshankawar** [comment]
@MaxYablochkin 
There's an old [issue / proposal](https://github.com/flutter/flutter/issues/17119) open to add search bar as a new widget that you can take a look and may share your thoughts if your proposal can be clubbed with that issue or you would want to keep this issue open and relate the other with this.

**Comment by MaxYablochkin** [comment]
@darshankawar 
I think this issue should be classified as a new one, as it relates to updating the specs to M3, but the implementation can probably be taken from the old issue #17119

**Comment by darshankawar** [comment]
Thanks for the feedback. Keeping this issue open and treating as proposal from M3 perspective.

**Comment by MaxYablochkin** [comment]
My attempt to render the SearchBar

![Screenshot_1673021519](https://user-images.githubusercontent.com/102767277/211053070-cadf7840-cf49-461e-a392-e4f9d7288995.png)
![Screenshot_1673021555](https://user-images.githubusercontent.com/102767277/211053073-014ad6b3-6a81-4ce7-a002-f4a33776fbe6.png)

```
import 'package:flutter/material.dart';

class SearchBar extends StatelessWidget {
  const SearchBar({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SafeArea(
        child: Align(
          alignment: Alignment.topCenter,
          child: InkWell(
            onTap: () {},
            radius: 120,
            splashColor: Theme.of(context).colorScheme.primary.withOpacity(0.2),
            highlightColor: Colors.transparent,
            borderRadius: BorderRadius.circular(30),
            child: Container(
              height: 56,
              width: 360,
              decoration: BoxDecoration(
                borderRadius: BorderRadius.circular(30),
                color: Theme.of(context)
                    .colorScheme
                    .primaryContainer
                    .withOpacity(0.5),
              ),
              child: Stack(
                alignment: Alignment.centerLeft,
                children: [
                  Padding(
                    padding: EdgeInsets.only(left: 3),
                    child: IconButton(
                      onPressed: () {},
                      icon: Icon(Icons.menu_outlined),
                    ),
                  ),
                  Padding(
                    padding: EdgeInsets.only(left: 60),
                    child: Text(
                      'Hinted search text',
                      style: Theme.of(context).textTheme.bodyLarge,
                    ),
                  ),
                  Padding(
                    padding: EdgeInsets.only(left: 310),
                    child: IconButton(
                      onPressed: () {},
                      icon: Icon(Icons.search),
                    ),
                  ),
                ],
              ),
            ),
          ),
        ),
      ),
    );
  }
}
```

**Comment by spiekermax** [comment]
Here is an implementation based on M3 specs with dynamic width:
```
class FloatingSearchBar extends StatelessWidget {
  const FloatingSearchBar({
    Key? key,
    this.height = 56,
    required this.leadingIcon,
    this.trailingIcon,
    required this.supportingText,
  }) : super(key: key);

  final double height;
  double get effectiveHeight {
    return max(height, 48);
  }

  final Widget leadingIcon;
  final Widget? trailingIcon;

  final String supportingText;

  @override
  Widget build(BuildContext context) {
    final ColorScheme colorScheme = Theme.of(context).colorScheme;
    final TextTheme textTheme = Theme.of(context).textTheme;

    return Padding(
      padding: const EdgeInsets.symmetric(horizontal: 16),
      child: Container(
        constraints: const BoxConstraints(minWidth: 360, maxWidth: 720),
        width: double.infinity,
        height: effectiveHeight,
        child: Material(
          elevation: 3,
          color: colorScheme.surface,
          shadowColor: colorScheme.shadow,
          surfaceTintColor: colorScheme.surfaceTint,
          borderRadius: BorderRadius.circular(effectiveHeight / 2),
          child: InkWell(
            onTap: () {},
            borderRadius: BorderRadius.circular(effectiveHeight / 2),
            highlightColor: Colors.transparent,
            splashFactory: InkRipple.splashFactory,
            child: Padding(
              padding: const EdgeInsets.symmetric(horizontal: 8),
              child: Row(children: [
                leadingIcon,
                Expanded(
                  child: Padding(
                    padding: const EdgeInsets.only(right: 8),
                    child: TextField(
                      cursorColor: colorScheme.primary,
                      style: textTheme.bodyLarge,
                      textAlignVertical: TextAlignVertical.center,
                      decoration: InputDecoration(
                        isCollapsed: true,
                        border: InputBorder.none,
                        contentPadding: const EdgeInsets.symmetric(horizontal: 8),
                        hintText: supportingText,
                        hintStyle: textTheme.bodyLarge?.apply(
                          color: colorScheme.onSurfaceVariant,
                        ),
                      ),
                    ),
                  ),
                ),
                if (trailingIcon != null) trailingIcon!,
              ]),
            ),
          ),
        ),
      ),
    );
  }
}
```

**Comment by SirusCodes** [comment]
This is resolved in #122309
